File: daily_problems/2026/07/0713/personal_submission/cf2206j_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, q;
	cin >> n >> q;

	vector<int> p1(n), p2(n), pos1(n), pos2(n);

	for (int i = 0; i < n; i ++) {
		cin >> p1[i]; p1[i] --;
		pos1[p1[i]] = i;
	}

	for (int i = 0; i < n; i ++) {
		cin >> p2[i]; p2[i] --;
		pos2[p2[i]] = i;
	}

	auto cost = [&] (int idx) -> int {
		if (idx <= 0 || idx >= n) return 0;
		int x = p2[idx - 1], y = p2[idx];
		int px = pos1[x], py = pos1[y];
		return (py - px + n) % n - 1;
	};

	auto calc = [&] (int idx1, int idx2) -> int {
		if (idx1 > idx2) swap(idx1, idx2);
		int ans = 0;
		ans += cost(idx1);
		ans += cost(idx1 + 1);
		if (idx1 + 1 != idx2) ans += cost(idx2);
		ans += cost(idx2 + 1);
		return ans;
	};

	long long ans = 0;
	for (int i = 1; i < n; i ++) ans += cost(i);

	q --;
	cout << ans + pos1[p2[0]] << '\n';

	while (q --) {
		int c, x, y;
		cin >> c >> x >> y;
		x --, y --;

		if (c == 1) {
			ans -= calc(pos2[p1[x]], pos2[p1[y]]);

			swap(p1[x], p1[y]);
			swap(pos1[p1[x]], pos1[p1[y]]);

			ans += calc(pos2[p1[x]], pos2[p1[y]]);
		}
		else {
			ans -= calc(x, y);

			swap(p2[x], p2[y]);
			swap(pos2[p2[x]], pos2[p2[y]]);

			ans += calc(x, y);
		}

		cout << ans + pos1[p2[0]] << '\n';
	}

	return 0;
}

File: daily_problems/2026/07/0713/personal_submission/cf2206j_yawn_sean.py
# Submission link: https://codeforces.com/contest/2206/submission/382175726
def main():
    n, q = MII()
    p1 = LGMI()
    p2 = LGMI()
    
    pos1 = [0] * n
    pos2 = [0] * n
    
    for i in range(n): pos1[p1[i]] = i
    for i in range(n): pos2[p2[i]] = i
    
    def cost(idx):
        if idx <= 0 or idx >= n: return 0
        x, y = p2[idx - 1], p2[idx]
        px = pos1[x]
        py = pos1[y]
        return (py - px) % n - 1
    
    def calc(idx1, idx2):
        if idx1 > idx2: idx1, idx2 = idx2, idx1
        ans = 0
        ans += cost(idx1)
        ans += cost(idx1 + 1)
        
        if idx1 + 1 != idx2:
            ans += cost(idx2)
        ans += cost(idx2 + 1)
        return ans
    
    ans = 0
    for i in range(1, n):
        ans += cost(i)
    
    outs = [ans + pos1[p2[0]]]
    
    for _ in range(q - 1):
        c, x, y = GMI()
        
        if c == 0:
            ans -= calc(pos2[p1[x]], pos2[p1[y]])
            
            p1[x], p1[y] = p1[y], p1[x]
            pos1[p1[x]], pos1[p1[y]] = pos1[p1[y]], pos1[p1[x]]
            
            ans += calc(pos2[p1[x]], pos2[p1[y]])
        
        else:
            ans -= calc(x, y)
            
            p2[x], p2[y] = p2[y], p2[x]
            pos2[p2[x]], pos2[p2[y]] = pos2[p2[y]], pos2[p2[x]]
            
            ans += calc(x, y)
        
        outs.append(ans + pos1[p2[0]])
    
    print('\n'.join(map(str, outs)))